# **Data Understanding**

This notebook loads the e-commerce dataset and inspects its structure, data types, summary statistics, missing values, and duplicate records.

In [24]:
# Load pandas for data manipulation.
import pandas as pd

# Set option to display all columns and format float values to two decimal places.
pd.set_option('display.max_columns', None)
pd.set_option('display.float_format', '{:,.2f}'.format)

In [25]:
# Read the raw e-commerce dataset.
ecommerce_df = pd.read_csv(r'dataset/raw/ecommerce_dataset_+1m.csv')

## 1. Look at the Data

In [26]:
# Display a random sample of 10 rows
ecommerce_df.sample(10)

,order_id,order_date,order_year,order_month,order_day,order_hour,order_minute,order_second,is_weekend,order_status,return_reason,customer_id,customer_name,gender,age,customer_segment,country,city,customer_loyalty_score,total_orders_by_customer,account_creation_date,product_id,product_name,category,sub_category,brand,product_rating_avg,product_reviews_count,stock_quantity,unit_price_usd,quantity,discount_percent,discount_amount_usd,total_price_usd,cost_usd,profit_usd,tax_usd,currency,payment_method,payment_status,installment_plan,shipping_method,shipping_cost_usd,delivery_days,shipping_country,warehouse_location,delivery_status,rating,review_sentiment,customer_feedback,coupon_used,coupon_code,campaign_source,device_type,traffic_source,session_duration_minutes,pages_visited,abandoned_cart_before,fraud_risk_score,profit_margin_percent,order_priority,support_ticket_created
229420,ORD-UIV6Z,2024-09-18 22:34:31.392627,2024,9,18,22,34,31,No,Completed,NaN,CUS-TCM1A,Jose Miller,Male,61,Regular,United States,Los Angeles,93.10,12,2021-04-25,PRD-O22Y,Treadmill Folding,Sports,Accessories,Lenovo,4.80,3457,147,95.72,1,5,4.79,90.93,45.26,45.67,5.20,USD,Apple Pay,Paid,No,Next Day,4.27,7,United States,USA-East,Delivered,3,Neutral,Average quality,Yes,CPN-1PJ,Facebook,Desktop,Search,34.10,20,No,58.00,50.23,Medium,Yes
739891,ORD-OVB7A,2024-03-15 04:30:51.039779,2024,3,15,4,30,51,No,Completed,NaN,CUS-28920,Nancy Smith,Female,46,Regular,Italy,Turin,47.00,42,2023-05-06,PRD-1HDK,Cotton T-Shirt Pack,Clothing,Kids Wear,Xiaomi,4.90,1591,453,99.51,2,25,49.76,149.26,114.88,34.38,13.02,USD,Debit Card,Paid,Yes,Next Day,1.45,2,Italy,UK,In Transit,2,Negative,Not satisfied,No,NaN,Affiliate,Mobile,Direct,14.80,6,Yes,75.60,23.03,Low,No
953927,ORD-6VCVR,2025-01-18 01:09:04.418240,2025,1,18,1,9,4,Yes,Processing,NaN,CUS-EY7D0,Christopher Hunt,Male,47,Premium,Germany,Munich,78.80,50,2024-01-04,PRD-QTYL,Running Shoes Air Zoom,Clothing,Shoes,Philips,4.30,4917,279,113.14,2,5,11.31,214.97,92.64,122.33,16.70,USD,Debit Card,Paid,No,Standard,12.73,3,Germany,UK,In Transit,5,Positive,Fast delivery,No,NaN,Affiliate,Mobile,Referral,60.00,15,Yes,29.40,56.91,High,Yes
970867,ORD-B9JT7,2025-12-16 10:32:50.751024,2025,12,16,10,32,50,No,Completed,NaN,CUS-KJ1CT,Mary Logan,Female,56,Regular,Germany,Munich,62.80,41,2022-10-19,PRD-TIUA,Denim Jeans Slim Fit,Clothing,Womens Wear,Philips,3.60,22,918,143.22,5,5,35.80,680.30,405.80,274.50,55.79,USD,Bank Transfer,Paid,No,Standard,1.01,14,Germany,UK,Delivered,2,Negative,Great product!,No,NaN,Facebook,Desktop,Direct,33.80,20,No,92.90,40.35,Low,No
916840,ORD-HCEOD,2024-04-04 20:00:23.944865,2024,4,4,20,0,23,No,Returned,Not As Described,CUS-C723V,Victoria York,Female,51,VIP,Germany,Berlin,12.50,9,2023-01-22,PRD-U775,Bedsheet Set Cotton,Home,Kitchen,Xiaomi,4.50,2453,409,41.60,5,20,41.60,166.40,128.80,37.60,13.96,USD,Debit Card,Paid,Yes,Express,15.64,10,Germany,USA-West,Delivered,1,Negative,Not satisfied,Yes,CPN-JLA,Affiliate,Mobile,Search,32.00,15,Yes,32.90,22.60,High,No
698420,ORD-WGE8E,2025-04-26 12:11:22.071161,2025,4,26,12,11,22,Yes,Completed,NaN,CUS-XWU21,Amanda Clark,Female,37,Regular,United Kingdom,Leeds,83.90,50,2022-09-03,PRD-GUK8,Swimming Goggles,Sports,Accessories,HP,3.20,2018,728,234.12,4,15,140.47,796.01,567.00,229.01,70.27,USD,Credit Card,Paid,Yes,Economy,20.36,14,United Kingdom,USA-East,Delivered,3,Neutral,Excellent!,Yes,CPN-W2F,Affiliate,Tablet,Referral,3.60,11,No,98.70,28.77,Medium,Yes
656212,ORD-M49V7,2024-06-08 00:03:23.476337,2024,6,8,0,3,23,Yes,Pending,NaN,CUS-9VZ9N,Jennifer Brown,Female,19,VIP,Italy,Rome,12.10,4,2021-10-22,PRD-AJL4,Dumbbells Set Adjustable,Sports,Gym Equipment,Anker,3.40,3686,811,116.13,1,0,0.00,116.13,63.86,52.27,9.66,USD,PayPal,Paid,Yes,Express,7.92,13,Italy,EU-Central,Delivered,3,Neutral,Excellent!,Yes,CPN-5NI,Google Ads,Mobile,Referral,56.30,20,Yes,20.90,45.01,Medium,No
768255,ORD-WKNDL,2024-10-14 12:28:32.687381,2024,10,14,12,28,32,No,Completed,NaN,CUS-QA4C3,Tara Nelson PhD,Female,65,Premium,Spain,Seville

In [27]:
# Dataset dimensions (rows, columns)
ecommerce_df.shape

(1000123, 62)

## 2. Data Types & Structure

In [28]:
# Concise summary of column types and non-null counts
ecommerce_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000123 entries, 0 to 1000122
Data columns (total 62 columns):
 #   Column                    Non-Null Count    Dtype  
---  ------                    --------------    -----  
 0   order_id                  1000123 non-null  object 
 1   order_date                1000123 non-null  object 
 2   order_year                1000123 non-null  int64  
 3   order_month               1000123 non-null  int64  
 4   order_day                 1000123 non-null  int64  
 5   order_hour                1000123 non-null  int64  
 6   order_minute              1000123 non-null  int64  
 7   order_second              1000123 non-null  int64  
 8   is_weekend                1000123 non-null  object 
 9   order_status              1000123 non-null  object 
 10  return_reason             99879 non-null    object 
 11  customer_id               1000123 non-null  object 
 12  customer_name             1000123 non-null  object 
 13  gender                    1

## 3. Descriptive Statistics

In [29]:
# Summary statistics for numerical columns
ecommerce_df.describe()

,order_year,order_month,order_day,order_hour,order_minute,order_second,age,customer_loyalty_score,total_orders_by_customer,product_rating_avg,product_reviews_count,stock_quantity,unit_price_usd,quantity,discount_percent,discount_amount_usd,total_price_usd,cost_usd,profit_usd,tax_usd,shipping_cost_usd,delivery_days,rating,session_duration_minutes,pages_visited,fraud_risk_score,profit_margin_percent
count,"1,000,123.00","1,000,123.00","1,000,123.00","1,000,123.00","1,000,123.00","1,000,123.00","1,000,123.00","1,000,123.00","1,000,123.00","1,000,123.00","1,000,123.00","1,000,123.00","1,000,123.00","1,000,123.00","1,000,123.00","1,000,123.00","1,000,123.00","1,000,123.00","1,000,123.00","1,000,123.00","1,000,123.00","1,000,123.00","1,000,123.00","1,000,123.00","1,000,123.00","1,000,123.00","1,000,123.00"
mean,"2,024.59",6.51,15.75,11.49,29.49,29.49,46.49,49.97,25.49,4.00,"2,501.57",499.81,147.02,3.00,8.50,37.48,403.24,242.31,160.93,30.25,12.50,7.51,3.00,30.51,10.51,49.96,39.47
std,0.58,3.45,8.80,6.92,17.31,17.33,16.73,28.89,14.43,0.58,"1,443.46",288.86,103.67,1.41,7.60,56.78,370.13,226.72,161.93,28.89,7.22,4.03,1.41,17.03,5.77,28.85,10.94
min,"2,024.00",1.00,1.00,0.00,0.00,0.00,18.00,0.00,1.00,3.00,0.00,0.00,10.00,1.00,0.00,0.00,7.53,4.03,0.70,0.43,0.00,1.00,1.00,1.00,1.00,0.00,6.66
25%,"2,024.00",4.00,8.00,5.00,14.00,14.00,32.00,24.90,13.00,3.50,"1,253.00",250.00,69.51,2.00,0.00,0.00,139.02,82.45,50.69,10.12,6.26,4.00,2.00,15.70,6.00,25.00,31.69
50%,"2,025.00",7.00,16.00,11.00,30.00,29.00,46.00,50.00,25.00,4.00,"2,503.00",500.00,125.11,3.00,10.00,16.91,290.19,172.53,108.79,21.28,12.50,8.00,3.00,30.50,11.00,50.00,39.90
75%,"2,025.00",10.00,23.00,17.00,44.00,44.00,61.00,75.00,38.00,4.50,"3,752.00",750.00,195.61,4.00,15.00,49.26,545.76,326.45,214.64,40.59,18.74,11.00,4.00,45.30,16.00,74.90,48.17
max,"2,026.00",12.00,31.00,23.00,59.00,59.00,75.00,100.00,50.00,5.00,"5,000.00","1,000.00",500.00,5.00,25.00,624.71,"2,499.75","1,743.70","1,495.75",245.59,25.00,14.00,5.00,60.00,20.00,100.00,60.02


## 4. Missing Values

In [30]:
# Percentage of missing values per column (%)
print((ecommerce_df.isnull().sum() / ecommerce_df.shape[0] * 100).to_string())

order_id                    0.00
order_date                  0.00
order_year                  0.00
order_month                 0.00
order_day                   0.00
order_hour                  0.00
order_minute                0.00
order_second                0.00
is_weekend                  0.00
order_status                0.00
return_reason              90.01
customer_id                 0.00
customer_name               0.00
gender                      0.00
age                         0.00
customer_segment            0.00
country                     0.00
city                        0.00
customer_loyalty_score      0.00
total_orders_by_customer    0.00
account_creation_date       0.00
product_id                  0.00
product_name                0.00
category                    0.00
sub_category                0.00
brand                       0.00
product_rating_avg          0.00
product_reviews_count       0.00
stock_quantity              0.00
unit_price_usd              0.00
quantity  

**Notable missing values:**
- `return_reason` → ~90.01% missing (expected, only present for returned orders)
- `customer_feedback` → ~19.96% missing
- `coupon_code` → ~50.00% missing (expected, only present when a coupon was used)

## 5. Duplicate Check

In [31]:
# Total number of fully duplicate rows
duplicates_row = ecommerce_df.duplicated().sum()
print(f'Total Duplicate Rows : {duplicates_row}')

Total Duplicate Rows : 0


## 6. Unique Identifier Check

In [32]:
"""
    Check which columns have completely unique values
    True = potential unique identifier
 """
print(ecommerce_df.apply(lambda col: col.is_unique).to_string())

order_id                    False
order_date                   True
order_year                  False
order_month                 False
order_day                   False
order_hour                  False
order_minute                False
order_second                False
is_weekend                  False
order_status                False
return_reason               False
customer_id                 False
customer_name               False
gender                      False
age                         False
customer_segment            False
country                     False
city                        False
customer_loyalty_score      False
total_orders_by_customer    False
account_creation_date       False
product_id                  False
product_name                False
category                    False
sub_category                False
brand                       False
product_rating_avg          False
product_reviews_count       False
stock_quantity              False
unit_price_usd

> **Observation:**  
> The dataset does **not** have a reliable unique identifier column.  
> Although `order_date` appears unique in this sample, timestamps are not dependable as primary keys in real-world scenarios (collisions can occur).